superset/viz.py

```diff
@@ -309,7 +309,8 @@ def get_df_payload(self, query_obj=None):
         stacktrace = None
         df = None
         cached_dttm = datetime.utcnow().isoformat().split('.')[0]
-        if cache_key and cache and not self.force:
+        can_use_cache = cache_key and cache and not self.force
+        if security_manager.has_cache_access(self.datasource) and can_use_cache:
```

Comment:
This logic should all reside in a single if statement, order from cheapest to most expensive, i.e., we should have `security_manager.has_cache_access(self.datasource)` evaluated last. 

As a side note I wonder if we should check the cache timeout here as well, i.e., for Druid datasources one would probably want a timeout of 0 which would imply don't cache.

Reply:
Also I feel deciphering what is going on with the cache is getting somewhat complicated. Maybe a class on connivence methods would help provide more clarity.